lib/mix/lib/mix/lock.ex

```diff
@@ -0,0 +1,238 @@
+        unlock(lock)
+        Process.delete(pdict_key)
+      end
+    end
+  end
+
+  defp lock(path) do
+    :ok = File.mkdir_p(path)
+
+    {:ok, socket} = :gen_tcp.listen(0, @listen_opts)
+    {:ok, port} = :inet.port(socket)
+
+    spawn_link(fn -> accept_loop(socket) end)
+
+    try_lock(path, socket, port)
+  end
+
+  defp try_lock(path, socket, port) do
+    port_path = Path.join(path, "port_#{port}")
+
+    :ok = File.write(port_path, <<port::unsigned-integer-32>>, [:raw])
+
+    case grab_lock(path, port_path, 0) do
+      {:ok, 0} ->
+        # We grabbed lock_0, so all good
+        %{socket: socket, path: path}
+
+      {:ok, _n} ->
+        # We grabbed lock_1+, so we need to replace lock_0 and clean up
+        take_over(path, port_path)
+        %{socket: socket, path: path}
+
+      {:taken, probe_socket} ->
+        # Another process has the lock, wait for close and start over
+        await_close(probe_socket)
+        try_lock(path, socket, port)
+
+      :invalidated ->
+        try_lock(path, socket, port)
+    end
+  end
+
+  defp grab_lock(path, port_path, n) do
+    lock_path = Path.join(path, "lock_#{n}")
+
+    case File.ln(port_path, lock_path) do
+      :ok ->
+        {:ok, n}
+
+      {:error, :eexist} ->
+        case probe(lock_path) do
+          {:ok, probe_socket} ->
+            {:taken, probe_socket}
+
+          :error ->
+            grab_lock(path, port_path, n + 1)
+        end
+
+      {:error, :enoent} ->
+        :invalidated
+    end
+  end
+
+  defp accept_loop(listen_socket) do
+    case :gen_tcp.accept(listen_socket) do
+      {:ok, socket} ->
+        _ = :gen_tcp.send(socket, @probe_data)
+        accept_loop(listen_socket)
+
+      {:error, reason} when reason in [:closed, :einval] ->
+        :ok
+    end
+  end
+
+  defp probe(port_path) do
+    with {:ok, <<port::unsigned-integer-32>>} when port > 0 <- File.read(port_path),
+         {:ok, socket} <- connect(port) do
+      case :gen_tcp.recv(socket, 0, @probe_timeout_ms) do
+        {:ok, @probe_data} ->
+          {:ok, socket}
+
+        {:error, _reason} ->
+          :gen_tcp.close(socket)
+          :error
+      end
+    else
+      _other -> :error
+    end
+  end
+
+  defp connect(port) do
+    # On Windows connecting to an unbound port takes a few seconds to
+    # fail, so instead we shortcut the check by attempting a listen,
+    # which succeeds or fails immediately
+    case :gen_tcp.listen(port, [reuseaddr: true] ++ @listen_opts) do
+      {:ok, socket} ->
+        :gen_tcp.close(socket)
+        # The port is free, so connecting would fail
+        {:error, :econnrefused}
+
+      {:error, _reason} ->
+        :gen_tcp.connect(@loopback, port, @connect_opts)
+    end
+  end
+
+  defp take_over(path, port_path) do
+    lock_path = Path.join(path, "lock_0")
+
+    # We linked to lock_N successfully, so port_path should exist
+    :ok = File.rename(port_path, lock_path)
+
+    {:ok, names} = File.ls(path)
```

Comment:
```suggestion
    names = File.ls!(path)
```

This applies to all the times you match on `:ok` or `{:ok, _}` when using `File` operations, so I'll stop commenting on those 😄